include/Controller.h

```diff
@@ -9,44 +9,36 @@
 #include <ctime>
 #include <iomanip>
 #include <functional>
+
 namespace csfoundation {
-    class Controller {
-    public:
-        Controller(bool verbose);
-        ~Controller();
-
-        void printInfo(std::string filename);
-    
-        template <class T> void printSample(std::vector<T> arr) {
-            for (auto& val : arr) {
-                std::cout << val << " ";
-            }
-            std::cout << std::endl;
-        }
-    
-        template <class T> double runAlgorithm(std::string const docfile, 
-            std::function<void(std::vector<T>)> func, 
-            std::vector<T> data) {
-                std::cout << "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - " << std::endl;
-            std::clock_t start;
-            double runtime;
-            this->printInfo(docfile);
-        
-            start = std::clock();
-            if (verbose) {
-                this->printSample(data);
-            }
-            func(data);
-            runtime = (std::clock() - start) / (double)CLOCKS_PER_SEC;
-            std::cout << std::fixed;
-            std::cout << "Sample size (" << data.size() << "): "  << std::setprecision(8)<< runtime 
-                << "s" << std::endl;
-            return runtime;
-        }
-
-    private:
-        bool verbose;
-
-    };
+  class Controller {
+  public:
+    Controller(bool verbose);
+    ~Controller();
+
+    void printInfo(std::string filename);
+
+    template <class T> void printSample(std::vector<T> arr) {
+      for (auto& val : arr) std::cout << val << " ";
+      std::cout << std::endl;
+    }
+
+    template <class T> void runAlgorithm(std::function<void(std::vector<T>)> func, std::vector<T> data) {
```

Comment:
@RenegadeTech, is it possible to define these functions in the associated cpp file (as with printInfo), or do the templating nature mean they need to remain here? Alternatively, would it be worth bringing the cpp file contents here instead?

(I want to avoid implementations scattered across multiple files as much as possible.)

Reply:
Due to the nature of templating, the definition is required to be in the same file.

You should still keep the source file because the makefile will not recompile the header file into a linkable object.

Reply:
Ah, makes sense. Thanks